diff -Nuar kunpengsecl-v1.0.0-pre/attestation/ras/config/config.go kunpengsecl-v1.0.0/attestation/ras/config/config.go
--- kunpengsecl-v1.0.0-pre/attestation/ras/config/config.go	2021-12-26 12:17:34.000000000 +0800
+++ kunpengsecl-v1.0.0/attestation/ras/config/config.go	2022-01-21 21:50:23.509081269 +0800
@@ -67,8 +67,8 @@
 	NullString            = ""
 	extKey                = ".key"
 	extCert               = ".crt"
-	RasRootKeyFileDefault = "./pca-root"
-	RasPcaKeyFileDefault  = "./pca-ek"
+	RasRootKeyFileDefault = strPathSysConf + "/default_test/pca-root"
+	RasPcaKeyFileDefault  = strPathSysConf + "/default_test/pca-ek"
 	RasRootPrivKeyFile    = "rasconfig.rootprivkeyfile"
 	RasRootKeyCertFile    = "rasconfig.rootkeycertfile"
 	RasPcaPrivKeyFile     = "rasconfig.pcaprivkeyfile"
@@ -97,13 +97,13 @@
 	RasExtRules           = "rasconfig.basevalue-extract-rules"
 	RasAutoUpdateConfig   = "rasconfig.auto-update-config"
 	RasAuthKeyFile        = "rasconfig.authkeyfile"
-	RasAuthKeyFileDefault = "./ecdsakey"
+	RasAuthKeyFileDefault = strPathSysConf + "/default_test/ecdsakey"
 	// RAC
 	RacIKeyCertFileDefault     = "./ic"
 	RacEKeyCertFile            = "racconfig.ekcert"
 	RacIKeyCertFile            = "racconfig.ikcert"
-	RacEKFileDefaultTest       = "./ectest"
-	RacIKeyCertFileDefaultTest = "./ictest"
+	RacEKFileDefaultTest       = strPathSysConf + "/default_test/ectest"
+	RacIKeyCertFileDefaultTest = strPathSysConf + "/default_test/ictest"
 	RacEKeyCertFileTest        = "racconfig.ekcerttest"
 	RacIKeyCertFileTest        = "racconfig.ikcerttest"
 	RacServer                  = "racconfig.server" // client connect to server
