diff -uNr kunpengsecl/attestation/rac/cmd/raagent/config.go kunpengsecl_1/attestation/rac/cmd/raagent/config.go
--- kunpengsecl/attestation/rac/cmd/raagent/config.go	2022-07-24 15:27:49.884111929 +0800
+++ kunpengsecl_1/attestation/rac/cmd/raagent/config.go	2022-07-24 21:23:06.296101035 +0800
@@ -53,14 +53,14 @@
 	crtExt             = ".crt"
 	ekCert             = "./ec"
 	ikCert             = "./ic"
-	ekCertTest         = "./ectest"
-	ikCertTest         = "./ictest"
+	ekCertTest         = "/etc/attestation/default_test/ectest"
+	ikCertTest         = "/etc/attestation/default_test/ictest"
 	confNullSeed       = -1
 	defaultTestMode    = false
 	defaultVerboseMode = false
 	defaultDigestAlg   = "sha1"
-	defaultImaLog      = "./ascii_runtime_measurements"
-	defaultBiosLog     = "./binary_bios_measurements"
+	defaultImaLog      = "/etc/attestation/default_test/ascii_runtime_measurements"
+	defaultBiosLog     = "/etc/attestation/default_test/binary_bios_measurements"
 	// ras server listen ip:port
 	lflagServer = "server"
 	sflagServer = "s"
diff -uNr kunpengsecl/attestation/rac/ractools/ractools.go kunpengsecl_1/attestation/rac/ractools/ractools.go
--- kunpengsecl/attestation/rac/ractools/ractools.go	2022-07-24 21:15:35.639281051 +0800
+++ kunpengsecl_1/attestation/rac/ractools/ractools.go	2022-07-24 21:38:57.170271072 +0800
@@ -112,8 +112,8 @@
 	Family: Not Applicable`
 
 	emptyPassword   = ""
-	TestImaLogPath  = "./ascii_runtime_measurements"
-	TestBiosLogPath = "./binary_bios_measurements"
+	TestImaLogPath  = "/etc/attestation/default_test/ascii_runtime_measurements"
+	TestBiosLogPath = "/etc/attestation/default_test/binary_bios_measurements"
 	TestSeedPath    = "./simulator_seed"
 	ImaLogPath      = "/sys/kernel/security/ima/ascii_runtime_measurements"
 	BiosLogPath     = "/sys/kernel/security/tpm0/binary_bios_measurements"
diff -uNr kunpengsecl/attestation/ras/cmd/config.yaml kunpengsecl_1/attestation/ras/cmd/config.yaml
--- kunpengsecl/attestation/ras/cmd/config.yaml	2022-07-24 15:27:49.884111929 +0800
+++ kunpengsecl_1/attestation/ras/cmd/config.yaml	2022-07-24 21:26:08.782052807 +0800
@@ -11,7 +11,7 @@
   hbduration: 10s
   trustduration: 2m0s
 rasconfig:
-  authkeyfile: ./ecdsakey.pub
+  authkeyfile: /etc/attestation/default_test/ecdsakey.pub
   pcakeycertfile: ""
   pcaprivkeyfile: ""
   httpsswitch: false
diff -uNr kunpengsecl/attestation/ras/config/config.go kunpengsecl_1/attestation/ras/config/config.go
--- kunpengsecl/attestation/ras/config/config.go	2022-07-24 15:27:49.884111929 +0800
+++ kunpengsecl_1/attestation/ras/config/config.go	2022-07-24 21:24:33.713035999 +0800
@@ -86,9 +86,9 @@
 	rasLogFile      = "./logs/ras-log.txt"
 	keyExt          = ".key"
 	crtExt          = ".crt"
-	rootKey         = "./pca-root"
-	eKey            = "./pca-ek"
-	authKey         = "./ecdsakey"
+	rootKey         = "/etc/attestation/default_test/pca-root"
+	eKey            = "/etc/attestation/default_test/pca-ek"
+	authKey         = "/etc/attestation/default_test/ecdsakey"
 	hbDuration      = 20   // seconds
 	trustDuration   = 1200 // seconds
 	digestAlgorithm = "sha1"
